[JENKINS-38187] Fix setting jnlp port through system property

Vlatombe opened the pull request:
https://issues.jenkins-ci.org/browse/JENKINS-38187

Fix setting up slave agent port through -Djenkins.model.Jenkins.slaveAgentPort
not only during first startup, but on any startup.

@reviewbybees
@jenkinsci/code-reviewers

Vlatombe committed:
[JENKINS-38187] Honor System Properties to set up jnlp in SetupWizard

Fix setting up slave agent port through -Djenkins.model.Jenkins.slaveAgentPort on startup.
Add -Djenkins.model.Jenkins.slaveAgentPortEnforce=true to enforce slave agent port on every start,  and an administrative monitor in case the value is modified through UI.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

Vlatombe commented:
Reviewing test failures

Vlatombe commented:
Not sure what is wrong with the remaining test. It passes locally for me.

oleg-nenashev commented:
I'm not sure about the change. I would not expect code modification here

daniel-beck commented:
I never liked this aspect of the original PR, as both possible solutions (ignore once configured despite being set on the commend line; and override on every restart while not completely disabling the UI option) have significant downsides.

Tentative :-1: but open to explanations why this needs to be done.

daniel-beck commented:
I would be extremely careful here: This looks like an effective regression for everyone skipping the wizard using system property.

Vlatombe commented:
I'm not specially attached to having the jnlp disabled directly in Jenkins. My primary focus is to fix setting up using `-Djenkins.model.Jenkins.slaveAgentPort` to avoid having to use hacks such as https://github.com/jenkinsci/docker/blob/master/init.groovy.

Vlatombe commented:
Pushed a second version without changing base behaviour of Jenkins.

Vlatombe commented:
@daniel-beck @oleg-nenashev wdyt now?

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -982,6 +982,7 @@ private Object readResolve() {
         if (jdks == null) {
             jdks = new ArrayList<>();
         }
+        slaveAgentPort = SystemProperties.getInteger(Jenkins.class.getName()+".slaveAgentPort", slaveAgentPort);
```
Still not sure about keeping the system property "active" beyond initial setup.

Vlatombe replied:
If the port is not enforced on each start, then people will still using 'hacks' such as https://github.com/jenkinsci/docker/blob/master/init.groovy, which is what I want to avoid.

I would actually go even further: if this property is set, the jnlp port shouldn't be editable through UI. Most cases this port is fixed because of firewall setup so you don't want anyone to change the value through UI once started.

daniel-beck replied:
Then this should have been the behavior from the start. Which [it is neither in behavior nor intent](https://github.com/jenkinsci/jenkins/commit/653fbdb65024b1b528e21f682172885f7111bba9#commitcomment-17342255), and it's [documented differently as well](https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties).

Or are you claiming that there is absolutely nobody who needed a fixed port for the initial setup, but changed it on the UI later?

The system property was unfortunately not named `initialPort` or similar due to @kohsuke circumventing the usual PR process, but it's too late to change that.

My proposal if you want to have different behavior, do it properly: New system property, called `enforcePort` (or similar), that removes the option from the UI, skips setting the option on form submission, and `readResolve`s.

KostyaSha replied:
> Most cases this port is fixed because of firewall setup so you don't want anyone to change the value through UI once started.

During debug (that issues constantly appears with security and firewalls) you may want to change it without restarting master. Inconsistency could be checked and reported to administrative monitor.

@daniel-beck i made own hacks for this port by shipping piece of `config.xml` in tests that applies to stock docker-container. 

PS. @kohsuke so would you use PRs next time?

daniel-beck reviewed (commented):
Looks much better now, although I have serious doubts about the readResolve.

It also changes the meaning of the system property. It was always only the initial state, to support automated setup, now it would be 'enforcing' on every restart.

Vlatombe commented:
I added a separate property to control enforcement (rather than a second property to set up a port value), and added an administrative monitor as suggested by @KostyaSha to warn in case the value is changed through UI.

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1206,6 +1219,42 @@ private void launchTcpSlaveAgentListener() throws IOException {
+            return Jenkins.getSlaveAgentPortInitialValue(slaveAgentPort);
+        }
+
+        public String getExpectedPortString() {
```
Why? Can't you use the int for the message?

Vlatombe replied:
Otherwise it will be formatted as `12 345` in the UI instead of 12345.

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1206,6 +1219,42 @@ private void launchTcpSlaveAgentListener() throws IOException {
+        }
+
+        /**
+         * Depending on whether the user said "yes" or "no", send him to the right place.
```
Lazy copy/paste with unnecessary code?

Vlatombe replied:
will remove

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1206,6 +1219,42 @@ private void launchTcpSlaveAgentListener() throws IOException {
         }
     }
 
+    @Extension
```
`@Restricted` missing

oleg-nenashev replied:
fixed

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/EnforceSlaveAgentPortAdministrativeMonitor/message.jelly:
```diff
@@ -0,0 +1,34 @@
+<!--
+The MIT License
+
+Copyright (c) 2016, CloudBees
```
`CloudBees, Inc.`

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/EnforceSlaveAgentPortAdministrativeMonitor/message.jelly:
```diff
@@ -0,0 +1,34 @@
+        ${%description(it.expectedPortString)}
+        <form method="post" action="${rootURL}/${it.url}/act" name="${it.id}">
+            <div style="float:right">
+                <f:submit name="fix" value="${%Fix}"/>
```
Should use a more descriptive label if it actually changes the value right then rather than navigates to a config page.

Vlatombe committed:
[JENKINS-38187] Fixes @daniel-beck reviews

Vlatombe commented:
@daniel-beck fixes done according to your last comments

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/EnforceSlaveAgentPortAdministrativeMonitor/message.properties:
```diff
@@ -1 +1,2 @@
-description=Slave Agent Port has been manually changed. Its value will be reset to {0} on restart.
+description=Slave Agent Port has been manually changed. Its value will be reset to {0,number,#} on restart.
+reset=Reset to {0,number,#}
```
*… now*; to emphasize what it does as opposed to what happens automatically as mentioned in the description preceding this button.

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/EnforceSlaveAgentPortAdministrativeMonitor/message.properties:
```diff
@@ -1 +1,2 @@
-description=Slave Agent Port has been manually changed. Its value will be reset to {0} on restart.
+description=Slave Agent Port has been manually changed. Its value will be reset to {0,number,#} on restart.
```
Should probably mention the system property that sets this, otherwise there's no reference to the reason for this or action to change this behavior.

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1234,15 +1235,8 @@ public int getExpectedPort() {
-         */
         public void doAct(StaplerRequest req, StaplerResponse rsp) throws IOException {
-            if (req.hasParameter("fix")) {
+            if (req.hasParameter("reset")) {
```
:ant: Still weird as there's no way to have a submission that doesn't enter this branch. There's no second button.

daniel-beck reviewed (approved):
Can be merged now I think, but I recommend minor changes to the messages.

Vlatombe committed:
[JENKINS-38187] Fix latest comments

oleg-nenashev commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -596,7 +597,16 @@ protected void onModified() throws IOException {
      * 0 for random, -1 to disable.
      */
-    private int slaveAgentPort = SystemProperties.getInteger(Jenkins.class.getName()+".slaveAgentPort",0);
+    private int slaveAgentPort = getSlaveAgentPortInitialValue(0);
```
Hmm. We were using -1 in SetupWizard as a default value. Is it just another inconsistency?

Vlatombe replied:
Well, using -1 was my first proposal, but it got rejected by Daniel because this was changing behaviour for people skipping setupWizard.

oleg-nenashev commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -596,7 +597,16 @@ protected void onModified() throws IOException {
+    /**
+     * If -Djenkins.model.Jenkins.slaveAgentPort is defined, enforce it on every start instead of only the first one.
+     */
+    private static final boolean SLAVE_AGENT_PORT_ENFORCE = SystemProperties.getBoolean(Jenkins.class.getName()+".slaveAgentPortEnforce", false);
```
🐛 From what I see it does not really enforce it since a Jenkins admin can actually change it. IMHO configuration flow needs to be adjusted:
* Values from the form submissions should be ignored
* Ideally the field should be readonly when this flag is set
* It should be possible forcing the disabled state (paranoid mode)

Vlatombe replied:
Thanks for the first bullet, I actually missed that comment from @daniel-beck earlier.

oleg-nenashev replied:
me too, sorry for duplication :(

oleg-nenashev commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1206,6 +1219,38 @@ private void launchTcpSlaveAgentListener() throws IOException {
+        @Override
+        public boolean isActivated() {
+            int slaveAgentPort = Jenkins.getInstance().slaveAgentPort;
+            return SLAVE_AGENT_PORT_ENFORCE && slaveAgentPort != Jenkins.getSlaveAgentPortInitialValue(slaveAgentPort);
```
Well, would be better to really enforce it

Vlatombe replied:
fixed

oleg-nenashev reviewed (changes requested):
🐛 since SLAVE_AGENT_PORT_ENFORCE flag does not really enforce the port. Dare admins can change it and break the instance

Vlatombe committed:
[JENKINS-38187] Make slaveAgentPort readonly in case the enforce flag is set

daniel-beck commented on core/src/main/java/hudson/security/GlobalSecurityConfiguration.java:
```diff
@@ -72,6 +72,10 @@ public int getSlaveAgentPort() {
         return Jenkins.getInstance().getSlaveAgentPort();
     }
 
+    public boolean isSlaveAgentPortEnforced() {
```
`@since @restricted`?

Vlatombe replied:
Done. Thanks for your continued feedback on this, I really appreciate it.

daniel-beck replied:
Nitpicking territory: We're not actually using `XXX`, at least not for `@since`, it's always `TODO`. Only change if there are further changes necessary anyway.

Vlatombe committed:
Add @since @restricted

Vlatombe committed:
s/XXX/TODO

daniel-beck commented:
FWIW I could have dealt with both approaches -- allowing to reconfigure temporarily (while showing a warning) can make sense in troubleshooting situations.

`EnforceSlaveAgentPortAdministrativeMonitor` still exists, I assume because of Groovy scripting?

Vlatombe commented:
@daniel-beck indeed, it is still possible to change the value through groovy scripting.

oleg-nenashev commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1104,11 +1104,21 @@ public int getSlaveAgentPort() {
     public void setSlaveAgentPort(int port) throws IOException {
+        if (!SLAVE_AGENT_PORT_ENFORCE) {
+            forceSetSlaveAgentPort(port);
+        }
```
NIT: else Logging or IOException

Vlatombe replied:
OK for logging, but I don't think an exception should be raised.

Vlatombe committed:
[JENKINS-38187] Log a warning message in case setSlaveAgentPort is called when the port is enforced

oleg-nenashev commented:
🐝
